module/os/linux/spl/spl-taskq.c

```diff
@@ -629,6 +778,10 @@ taskq_dispatch(taskq_t *tq, task_func_t func, void *arg, uint_t flags)
 
 	wake_up(&tq->tq_work_waitq);
 
+	TQSTAT_INC(tq, tasks_dispatched);
+
+	t->tqent_time[TQENT_TIME_QUEUED] = gethrtime();
```

Comment:
I've started in times when getting times in such a busy contexts and just for statistics was prohibitively expensive.  It should still be possible to find some Core2Duo system where the best timer available is HPET or ACPI timer in chipset.  I haven't looked much on this patch yet, but this point makes me worry of hos expensive would it be to have this always running on every system.

Reply:
Yeah, it bothered me while putting this together too. I've had a quick look through the kernel docs and I think the jiffy clock is probably good enough for this. We only care about relative time between the stages, and really we want to make large changes in behaviour visible; fine detail doesn't matter. I'll try it later today.

Reply:
So by default, 1 jiffy == 4 milliseconds, which is rather larger than I would have liked. I did a couple of tests with it, and it renders min/max and 10/100 useless, and 1000 iffy. 10000 is still not so bad.

I did have a bit of a read of the kernel timekeeping code (twisty!) and I didn't find anything there doesn't incur a hardware hit, other than `coarse`, which is ultimately just at the jiffy frequency.

Seems like our options are hi-res all the time, jiffies all the time, or some kind of in-between. Ideas (add your own):

- jiffies by default, with a tuneable to change to hi-res. Maybe useful; you see something that looks off, or you're trying to debug, so you flip the switch, have a look, turn it off when you're done
- hi-res by default, with a tuneable to change it to jiffies if its causing problems. I don't like this at all, mostly because its hard to know if its causing problems and not easy to discover
- self-tuning, switch to between based on ... something (load? clock overhead? step change in throughput? (handwaving details))
- hi-res, but only sample every nth task, or only sample at task start and ignore the task if we've done one "recently"

I'm kind of out my depth here; I don't know what I don't know. I'd be content to just go with jiffies for now, or remove the timing stuff altogether (timing is a separate commit for a reason!).

Reply:
Agreed, this is probably too expensive to be the default.  Alternately, we could add this hi-res timing code and then not enabled it by default.  This would make it available for profiling as needed and avoid incurring this cost on all system until we can better quantify what the performance impact would be.

I'm not against having a jiffie based approach available but given how coarse that timing is I kind of suspect it may only help uncover things which are way off.    Even for kernels which set CONFIG_HZ=1000 by default it's still quite rough.

Reply:
I think for the moment I will just drop the timing stuff, but keep the counters. I'm not really satisfied with the way the rolling averages are done either, and I have some other ideas that I'm playing with. We can do something better in the future if we figure it out.

Reply:
That sounds good.